textworld/logic/__init__.py

```diff
@@ -440,8 +489,11 @@ def multi_subtypes(self, types: Collection[Type]) -> Iterable[Collection[Type]]:
         Computes the descendant closure of a sequence of types, including the
         initial types.
         """
-        yield tuple(types)
-        yield from self.multi_descendants(types)
+        types = tuple(types)
+        if types not in self._cache:
+            self._cache[types] = [types] + list(self.multi_descendants(types))
+
+        yield from self._cache[types]
```

Comment:
There's no reason for this to be a generator any more since it's not lazy, so let's just cache `tuple`s and return them directly.

Reply:
Great.